### ISSUE TITLE
butterflypack-2.2.0 build failure with intel-19 in CI

### ISSUE THREAD

**Issue by balay** [xsdk-project/xsdk-issues#180]
```
55    /data/xsdkci/VS1mJQ1K/1/xsdk-project/spack-xsdk/spack-stage/balay/s
            pack-stage-butterflypack-2.2.0-6u7novzxda2qwljvvngrfu7rg3fparhz/spa
            ck-src/SRC_SINGLE/sMAGMA_utilities.f90(280): error #5082: Syntax er
            ror, found ',' when expecting one of: ( : % [ . = =>
     156            CBIND_DT,  value :: alpha, beta
     157    ----------------^
```
etc..

ref: https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3102857097

$ nice ./bin/spack install -j24 xsdk@0.8.0 %intel@19.1.1.217 ^netlib-lapack

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/9676872/spack-build-out.txt)

**Comment by liuyangzhuan** [comment]
@balay I don't see this when using gcc compilers. Does spack install butterflypack@2.1.1%intel@19.1.1.217 ^netlib-lapack work?

**Comment by liuyangzhuan** [comment]
@balay this is due to a recent commit that somehow fails the detection of intel compilers in cmake, I'm packing a v2.2.1 to fix this. Can you approve this? 
https://github.com/spack/spack/pull/32908
I will also udpate the xsdk sheet with v2.2.1

**Comment by balay** [comment]
> Does spack install butterflypack@2.1.1%intel@19.1.1.217 ^netlib-lapack work

Yes.
```
balay@xsdk:/data/balay/spack.x>./bin/spack find  butterflypack
==> 1 installed package
-- linux-centos7-cascadelake / intel@19.1.1.217 -----------------
butterflypack@2.1.1
```

**Comment by balay** [comment]
butterflypack@2.2.1 build now goes through with  intel@19.1.1.217

```
balay@xsdk:/data/balay/spack.x>./bin/spack find butterflypack
==> 2 installed packages
-- linux-centos7-cascadelake / intel@19.1.1.217 -----------------
butterflypack@2.1.1  butterflypack@2.2.1
```

**Comment by liuyangzhuan** [comment]
Thank you!

**Comment by balay** [comment]
@liuyangzhuan I'm seeing a couple of failures with butterflypack@2.2.1 [filing them here - instead of opening a new issue]

- linux-ubuntu20.04-skylake / oneapi@2022.2.0

```
  >> 510    ld: ../SRC_DOUBLECOMPLEX/libzbutterflypack.so.2.2.1: undefined refe
            rence to `__kmpc_end_masked'
  >> 511    ld: ../SRC_DOUBLECOMPLEX/libzbutterflypack.so.2.2.1: undefined refe
            rence to `__atomic_compare_exchange'
  >> 512    ld: ../SRC_DOUBLECOMPLEX/libzbutterflypack.so.2.2.1: undefined refe
            rence to `__atomic_load'
  >> 513    ld: ../SRC_DOUBLECOMPLEX/libzbutterflypack.so.2.2.1: undefined refe
            rence to `__kmpc_masked'
```
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3108965159

[spack-build-out(3).txt](https://github.com/xsdk-project/xsdk-issues/files/9685670/spack-build-out.3.txt)

- darwin-catalina-skylake / apple-clang@12.0.0

```
>> 2345    CMake Error at SRC_DOUBLE/cmake_install.cmake:93 (file):
     2346      file INSTALL cannot find
     2347      "/Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/spack-sta
             ge/spack-stage-butterflypack-2.2.1-jeta6ai2hih43v5b45k3boimoudhkpn
             r/spack-src/SRC_DOUBLE/ButterflyPACK_config.findButterflyPACK_conf
             ig.fi":
     2348      No such file or directory.
     2349    Call Stack (most recent call first):
     2350      cmake_install.cmake:73 (include)
```
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3108965071

[spack-build-out(5).txt](https://github.com/xsdk-project/xsdk-issues/files/9685672/spack-build-out.5.txt)

**Comment by liuyangzhuan** [comment]
@balay I tried oneapi on jlse and I don't see any error. Note that I'm using the following for 
	-DTPL_BLAS_LIBRARIES="$MKLROOT/lib/intel64/libmkl_intel_lp64.so;$MKLROOT/lib/intel64/libmkl_sequential.so;$MKLROOT/lib/intel64/libmkl_core.so" \
-DTPL_LAPACK_LIBRARIES="$MKLROOT/lib/intel64/libmkl_intel_lp64.so;$MKLROOT/lib/intel64/libmkl_sequential.so;$MK
LROOT/lib/intel64/libmkl_core.so" \

But from your build log:
TPL_BLAS_LIBRARIES='/opt/intel/oneapi/mkl/2022.2.0/lib/intel64/libmkl_intel_lp64.so;/opt/intel/oneapi/mkl/2022.2.0/lib/intel64/libmkl_sequential.so;/opt/intel/oneapi/mkl/2022.2.0/lib/intel64/libmkl_core.so;/usr/lib/x86_64-linux-gnu/libpthread.so;/usr/lib/x86_64-linux-gnu/libm.so;/usr/lib/x86_64-linux-gnu/libdl.so'

The use of libpthread looks suspicious. Can you remove libpthread (and perhaps also libdl) and try it again?

**Comment by liuyangzhuan** [comment]
@balay I haven't got any feedback on this. Have you tried to remove libpthread and libdl?

**Comment by balay** [comment]
@liuyangzhuan my build is via spack - so I'm not sure how that can be done.

BTW: I think different version of Intel compilers behave differently :(

For this - I can work-around by switching just butterflypack to gcc  in CI [and keep the remaining xsdk pkgs on intel compiler]. But would be good if the intel build works..

BTW: Latest failure at

https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3161006323

**Comment by balay** [comment]
@liuyangzhuan I see MacOS failure with butterflypack

https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3161006279

```
>> 2341    error: /Library/Developer/CommandLineTools/usr/bin/install_name_to
             ol: for: /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt
             /spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.
             2.1-xmldbsttvje74gwepvdrbwrw3ujoi663/lib/libdbutterflypack.2.2.1.d
             ylib (for architecture x86_64) option "-add_rpath /Users/xsdk/buil
             ds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spack/darwin-catalina-sk
             ylake/apple-clang-12.0.0/butterflypack-2.2.1-xmldbsttvje74gwepvdrb
             wrw3ujoi663/lib" would duplicate path, file already has LC_RPATH f
             or: /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spac
             k/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.1-x
             mldbsttvje74gwepvdrbwrw3ujoi663/lib
...
```
[spack-build-out(4).txt](https://github.com/xsdk-project/xsdk-issues/files/9770992/spack-build-out.4.txt)

**Comment by liuyangzhuan** [comment]
Hi, @balay Sorry for the delay on this. I made a few corrections:
(1) Now if you pass cmake with -Denable_openmp=OFF, the linking issues regarding omp should be gone for intel compilers, that will cover 
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3108965159
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3161006323

(2) The error ""/Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/spack-sta
             ge/spack-stage-butterflypack-2.2.1-jeta6ai2hih43v5b45k3boimoudhkpn
             r/spack-src/SRC_DOUBLE/ButterflyPACK_config.findButterflyPACK_conf
             ig.fi":" in
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3108965071
https://gitlab.com/xsdk-project/spack-xsdk/-/jobs/3161006279
should be fixed now. 

Can you give it a try using the master branch of butterflypack? If it works, I can pack a version 2.2.2.

**Comment by v-dobrev** [comment]
Just FYI, building `butterflypack@master` with gcc 6.3.1 is failing for me with errors like this:
```console
/tmp/dobrev1/spack-stage/spack-stage-butterflypack-master-zagz2b7pkpwqosjkudnnxzzmhfz3nbow/spack-src/SRC_DOUBLE/dBplus_utilities.f90:6483:27:

                      !$omp taskloop default(shared) private(j,rank,nn,flop,index_j,index_j_loc_s)
                           1
Error: Unclassifiable OpenMP directive at (1)
```
Is there a minimal version of gcc that is required?

**Comment by liuyangzhuan** [comment]
@v-dobrev Yes, I think the gcc version should be at least gcc7, we have seen this before 
https://github.com/xsdk-project/xsdk-issues/issues/127

Maybe I should add a minimum gcc version requirement in spack

**Comment by v-dobrev** [comment]
> @v-dobrev Yes, I think the gcc version should be at least gcc7, we have seen this before #127

Ah, I see, thanks! I didn't see that issue.

> Maybe I should add a minimum gcc version requirement in spack

Good idea -- that will be helpful.

**Comment by liuyangzhuan** [comment]
@balay Do you see further failures? Should I pack a version 2.2.2?

**Comment by balay** [comment]
sorry missed this one. Will start a test.

>  Now if you pass cmake with -Denable_openmp=OFF

How do we set this up in spack?

**Comment by balay** [comment]
attempting to add `butterflypack+openmp` variant - will see how that goes...

**Comment by liuyangzhuan** [comment]
@balay Thank you! Maybe you could also add a requirement of minimum version 7.0.0 for gnu compilers?

**Comment by balay** [comment]
pushed a change to xsdk branch. Will have to make a separate PR of all these changes directly to spack later [after more testing - and perhaps when 2.2.2 is released]

**Comment by balay** [comment]
I now get (on Mac):
```
  >> 2341    error: /Library/Developer/CommandLineTools/usr/bin/install_name_tool: for: /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2
             .2.2-ib2exmpdv4hzckotsx4bxknonww7o2sc/lib/libdbutterflypack.2.2.1.dylib (for architecture x86_64) option "-add_rpath /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spack/darwin-catalina-
             skylake/apple-clang-12.0.0/butterflypack-2.2.2-ib2exmpdv4hzckotsx4bxknonww7o2sc/lib" would duplicate path, file already has LC_RPATH for: /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/s
             pack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-ib2exmpdv4hzckotsx4bxknonww7o2sc/lib
```

```
xsdk@ypro spack-xsdk % ls /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-ib2exmpdv4hzckotsx4bxknonww7o2sc                         
ls: /Users/xsdk/builds/cYLkqamb/0/xsdk-project/spack-xsdk/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-ib2exmpdv4hzckotsx4bxknonww7o2sc: No such file or directory
```

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/9816369/spack-build-out.txt)

**Comment by balay** [comment]
Ah I had a typo with my 'butterflypack~openmp' code. Fixed and retesting.

OneAPI build goes through now.

**Comment by balay** [comment]
MacOS failure persists

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/9822893/spack-build-out.txt)

**Comment by liuyangzhuan** [comment]
There are two errors from the log:
1. error: /Library/Developer/CommandLineTools/usr/bin/install_name_tool: for: /Users/xsdk/spack/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-zoec4usndr32oy4sqogaa2bezswvjb7c/lib/libdbutterflypack.2.2.1.dylib (for architecture x86_64) option "-add_rpath /Users/xsdk/spack/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/netlib-lapack-3.9.1-tg2a5aa5wvwym5mzc5em2cnpiy5frapf/lib" would duplicate path, file already has LC_RPATH for: /Users/xsdk/spack/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/netlib-lapack-3.9.1-tg2a5aa5wvwym5mzc5em2cnpiy5frapf/lib
2. CMake Error at SRC_DOUBLE/cmake_install.cmake:93 (file):
  file INSTALL cannot find
  "/Users/xsdk/spack/spack-stage/spack-stage-butterflypack-2.2.2-zoec4usndr32oy4sqogaa2bezswvjb7c/spack-src/SRC_DOUBLE/ButterflyPACK_config.fin":
  No such file or directory.

I don't know what's the cause of the first one. But for the second one it's a failure of sed. I'm wondering what sed (GNU or BSD? Which version?) are you using on this macbook?

**Comment by balay** [comment]
Its macos native sed:
```
SED(1)                    BSD General Commands Manual                   SED(1)
```
Ok trying with:
```
+    depends_on("sed", type="build")
```

ok the build goes through now.

```
==> Installing butterflypack-2.2.2-qtep5reegrvb2xbxvqweoyivwbviwdsa
==> No binary for butterflypack-2.2.2-qtep5reegrvb2xbxvqweoyivwbviwdsa found: installing from source
==> No patches needed for butterflypack
==> butterflypack: Executing phase: 'cmake'
==> butterflypack: Executing phase: 'build'
==> butterflypack: Executing phase: 'install'
==> butterflypack: Successfully installed butterflypack-2.2.2-qtep5reegrvb2xbxvqweoyivwbviwdsa
  Fetch: 1m 23.60s.  Build: 2m 3.81s.  Total: 3m 27.41s.
[+] /Users/xsdk/spack/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-qtep5reegrvb2xbxvqweoyivwbviwdsa

```

**Comment by liuyangzhuan** [comment]
That's fantastic! Are you committing these changes in butterflypack.py to spack? If not, can you send me a copy of butterflypack.py and I will create a PR.

**Comment by liuyangzhuan** [comment]
Of course when there is no more issue in xsdk.

**Comment by balay** [comment]
Ok - PR at https://github.com/spack/spack/pull/33416

**Comment by liuyangzhuan** [comment]
From "/Users/xsdk/spack/opt/spack/darwin-catalina-skylake/apple-clang-12.0.0/butterflypack-2.2.2-qtep5reegrvb2xbxvqweoyivwbviwdsa" it seems that you created a version 2.2.2 for butterflypack? But I don't see it from 
https://github.com/spack/spack/pull/33416

I'm confused how you did it. Do you want me to create a version 2.2.2 for the xsdk release?

**Comment by balay** [comment]
Since 2.2.2 is not yet released - I removed that change from the PR.

We still haven't done much LCF testing - don't know if they will show additional issues. If you want to spin 2.2,2 now - and  additional releases (or patches)  for any new issues that might come up - that's fine with me.

**Comment by liuyangzhuan** [comment]
Satish, if you can work with 2.2.1 for this xsdk release, I don't need to pack newer versions.

**Comment by balay** [comment]
My testing was with 2.2.2 - I think this openmp change is in 2.2.2? If so - we might need 2.2.2 for this MR?

Sorry wasn't thinking about this issue..

For now - will mark it as draft.

**Comment by balos1** [comment]
This seems to be fixed now with 2.2.2, so closing.
